allennlp/models/encoder_decoders/encoder_decoder.py

```diff
@@ -0,0 +1,320 @@
+        normalization.
+
+        Parameters
+        ----------
+        target_scores : torch.LongTensor
+            Unnormalized class scores, the output from a projection layer. (batch_size, num_classes)
+
+        Returns
+        -------
+        prediction_dict : Dict[str, torch.LongTensor] with fields
+            class_probabilities
+            predicted_classes
+        """
+        # (batch_size, num_classes)
+        class_probabilities = F.softmax(target_scores)
+        # (batch_size,)
+        _, predicted_classes = torch.max(class_probabilities, 1)
+        prediction_dict = {"class_probabilities": class_probabilities, "predicted_classes": predicted_classes}
+        return prediction_dict
+
+    @staticmethod
+    def _get_loss(logits: torch.LongTensor,
+                  targets: torch.LongTensor,
+                  target_mask: torch.LongTensor) -> torch.LongTensor:
+        """
+        Takes logits (unnormalized outputs from the decoder) of size (batch_size, num_decoding_steps - 1,
+        num_classes), target indices of size (batch_size, num_decoding_steps) and corresponding masks of size
+        (batch_size, num_decoding) steps and computes cross entropy loss while taking the mask into account.
+
+        The length of ``targets`` is expected to be greater than that of ``logits`` because the decoder does
+        not need to compute the output corresponding to the last timestep of ``targets``. This method aligns
+        the inputs appropriately to compute the loss.
+        """
+        # `targets` currently contains START + target_sequence + END, and logits correspond to the outputs
+        # from these timesteps till the second from last.
+        # During training, we want the output from timestep i to be similar to the
+        # input token from timestep i + 1. That is, the targets should be shifted by one timestep for
+        # appropriate comparison.
+        # Consider a single example where the target has 3 words, and padding is to 7 tokens. The complete
+        #   sequence would correspond to <S> w1 w2 w3 <E> <P> <P>
+        #   and the mask would be         1  1  1  1   1   0   0
+        #   and let the logits be         l1 l2 l3 l4  l5  l6
+        # We actually need to compare:
+        #   the sequence w1 w2 w3 <E> <P> <P>
+        #   with masks   1  1  1   1   0   0
+        #   against      l1 l2 l3  l4  l5  l6
+        relevant_targets = targets[:, 1:].contiguous()  # (batch_size, num_decoding_steps - 1)
+        relevant_mask = target_mask[:, 1:].contiguous()  # (batch_size, num_decoding_steps - 1)
+        loss = sequence_cross_entropy_with_logits(logits, relevant_targets, relevant_mask)
+        return loss
+
+    @overrides
+    def decode(self, output_dict: Dict[str, torch.Tensor]) -> Dict[str, torch.Tensor]:
+        """
+        Trims the output predictions to the first end symbol, replaces indics with corresponding tokens,
```

Comment:
The name of this method clashes with the "decoder" of `EncoderDecoder`.  You implemented this correctly, but I think a reader could easily get confused about what this method does, given its name.  It'd be helpful to have a comment in the docstring that says something like "`Model.decode` is for doing inference or finalizing predictions after a call to `Model.forward`, and is separate from the notion of "decoding" in our `EncoderDecoder` model.  That decoder lives in `forward`; this `decode` method just cleans up the output and converts it to strings."